driver/src/main/com/mongodb/GridFSBucketImpl.java

```diff
@@ -0,0 +1,316 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import static com.mongodb.ReadPreference.primary;
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.lang.String.format;
+
+final class GridFSBucketImpl implements GridFSBucket {
+    private final MongoDatabase database;
+    private final String bucketName;
+    private final int chunkSizeBytes;
+    private final CodecRegistry codecRegistry;
+    private final WriteConcern writeConcern;
+    private final ReadPreference readPreference;
+    private final MongoCollection<BsonDocument> filesCollection;
+    private final MongoCollection<BsonDocument> chunksCollection;
+    private boolean checkedIndexes;
+
+    GridFSBucketImpl(final MongoDatabase database) {
+        this(database, "fs");
+    }
+
+    GridFSBucketImpl(final MongoDatabase database, final String bucketName) {
+        this.database = notNull("database", database);
+        this.bucketName = notNull("bucketName", bucketName);
+        this.chunkSizeBytes = 255;
+        this.codecRegistry = database.getCodecRegistry();
+        this.writeConcern = database.getWriteConcern();
+        this.readPreference = database.getReadPreference();
+        this.filesCollection = getFilesCollection(BsonDocument.class);
+        this.chunksCollection = getChunksCollection(BsonDocument.class);
+    }
+
+    GridFSBucketImpl(final MongoDatabase database, final String bucketName, final int chunkSizeBytes, final CodecRegistry codecRegistry,
+                     final ReadPreference readPreference, final WriteConcern writeConcern,
+                     final MongoCollection<BsonDocument> filesCollection, final MongoCollection<BsonDocument> chunksCollection,
+                     final boolean checkedIndexes) {
+        this.database = notNull("database", database);
+        this.bucketName = notNull("bucketName", bucketName);
+        this.chunkSizeBytes = chunkSizeBytes;
+        this.codecRegistry = notNull("codecRegistry", codecRegistry);
+        this.readPreference = notNull("readPreference", readPreference);
+        this.writeConcern = notNull("writeConcern", writeConcern);
+        this.checkedIndexes = checkedIndexes;
+        this.filesCollection =  notNull("filesCollection", filesCollection);
+        this.chunksCollection = notNull("chunksCollection", chunksCollection);
+    }
+
+    @Override
+    public String getBucketName() {
+        return bucketName;
+    }
+
+    @Override
+    public int getChunkSizeBytes() {
+        return chunkSizeBytes;
+    }
+
+    @Override
+    public CodecRegistry getCodecRegistry() {
+        return codecRegistry;
+    }
+
+    @Override
+    public ReadPreference getReadPreference() {
+        return readPreference;
+    }
+
+    @Override
+    public WriteConcern getWriteConcern() {
+        return writeConcern;
+    }
+
+    @Override
+    public GridFSBucket withBucketName(final String bucketName) {
```

Comment:
I'd prefer that we leave this out.  We don't have  

    MongoDatabase withName(String name)

on MongoDatabase.